lib/AST/Printer/Printer.cpp

```diff
@@ -596,33 +660,47 @@ TokenPrinterContext::~TokenPrinterContext(void) {
 }
 
 // More typical APIs when we've got PASTA ASTs.
-PrintedTokenRange PrintedTokenRange::Create(const Decl &decl) {
-  return PrintedTokenRange::Create(
-      decl.ast, const_cast<clang::Decl *>(decl.u.Decl));
+PrintedTokenRange PrintedTokenRange::Create(
+    const Decl &decl, const PrintingPolicy &pp, bool maintain_provenance) {
+  auto ret = PrintedTokenRange::Create(
+      decl.ast, const_cast<clang::Decl *>(decl.u.Decl), pp);
+  if (!maintain_provenance) {
+    ret.DumpProvenanceInformation();
+  }
+  return ret;
 }
 
 // More typical APIs when we've got PASTA ASTs.
-PrintedTokenRange PrintedTokenRange::Create(const Stmt &stmt) {
-  return PrintedTokenRange::Create(
-      stmt.ast, const_cast<clang::Stmt *>(stmt.u.Stmt));
+PrintedTokenRange PrintedTokenRange::Create(
+    const Stmt &stmt, const PrintingPolicy &pp, bool maintain_provenance) {
+  auto ret = PrintedTokenRange::Create(
```

Comment:
I've removed the `maintain_provenance` parameter, as the `DumpProvenanceInformation` API is public anyway. Besides this, what copying and pasting do you want eliminated? The types differ, and private fields of `Decl`, `Stmt`, and `Type` are accessed and differently-casted, and the public header only forward declares these classes.

Reply:
I was aiming at (u have similar thing right below)
```
auto ret = PrintedTokenRange( x.ast, y, pp )
if (!maintain_provenance)
   ret.DumpProvenanceInformation();
return ret;
```

which can be `return foo( x, y, maintain )` and on `x` you will call `.ast` inside the `foo` (or just pass in `x.ast`) and it will implement the snippet above.
I am not sure if this exact suggestion is applicable at other places, but here it should?